### PR TITLE
HIP simulation in SiStrips

### PR DESCRIPTION
**Pull request by sdevissc**
This PR is for now _not_ meant to be merge now. We want to give the possibility to anybody to easily look and test the code.
## Some explanation about how it works:

a) The file SimTracker/SiStripDigitizer/data/APVProbaList.txt contains one line per detID (module). For each detId (first colum), a probability to have a HIP is given (second column). 

b) The textfile is read in the digitizer before looping on the event (filling a std::map <int, float>)

c) When looping over the events and looping over the simHits (and therefore over detId), the map is read and a random number is drawn and compared with the corresponding entry in the map. If the number is smaller than the entry, the corresponding _APV_ is flagged as bad (i.e. detAmpl are then set to zero), otherwise nothing happens.
## Some remarks and further details

A) By default, the HIP simulation is set to False in SimGeneral/MixingModule/python/SiStripSimParameters_cfi.py , see APVSaturationFromHIP, don't forget to change it to True to test this code.

B) About the map of probabilities: this is where most of the work is done. The code is available here: 

https://github.com/cms-sw/cmssw/compare/CMSSW_8_1_X...sdevissc:ProbaMapProducer?expand=1
 This map producer is for now not supposed to be merged, but later on, it might be interesting to have this option in the digitizer, so that new maps can be generated easily and without having to maintain an additional code (if we keep the map generation as it is now).

C) The map is for now computed using MinBias only, i.e. using a neutrino gun, with the PU scenario  AVE_35_BX_25ns. 

cmsDriver.py SingleNuE10_cfi.py  --conditions auto:run2_mc --pileup_input das:/RelValMinBias_13/CMSSW_8_1_0_pre2-80X_mcRun2_asymptotic_v10_gs810pre2-v1/GEN-SIM -n $1 --era Run2_2016 --eventcontent FEVTDEBUGHLT -s GEN,SIM,DIGI:pdigi_valid,L1,DIGI2RAW,HLT:@relval2016 --datatier GEN-SIM-DIGI-RAW-HLTDEBUG --pileup AVE_35_BX_25ns --filein das:/RelValTTbar_13/CMSSW_8_1_0_pre2-80X_mcRun2_asymptotic_v10_gs810pre2-v1/GEN-SIM ...

This choice constitute hopefully a reasonable starting point.

D) The current simulation does NOT include the modification of the cluster charges: the APV are ON or OFF. Affecting the charge is the next step of the algorithm development, and explains why I have already put in the code a new vector<bool>  called hipChannels and a new map<int, vector<bool>> called allHIPChannels. The idea behind is to ultimately separate the APVs known to be dead, and those affected by APV where we have a short dead time and then a recovery time characterized by a wrong estimation of the cluster charge.

E) For now, the only parameter to tune the simulation when starting a production (i.e. this does not concerns the map generation) is APVSaturationProbScaling, it multiplies the probability read in the map by its value.

F) as an illustration, with the current configuration on gets the following global hit efficiencies in the MultiTrackValidator (using RelVal TTbar as input):

capture d ecran 2016-06-29 a 17 28 10

There are clearly differences w.r.t. what can be observed in the data, see e.g. https://indico.cern.ch/event/535777/contributions/2200584/attachments/1288176/1917381/HipMitigation.pdf . Different sources for these discrepancies are possible:
- PU scenario (and reweighting)
- Absence of cluster charge change in HIP-affected APVs
- ...

G) With this method, we start from the simHits. This means that we start from cases where the interaction between the simulated particle and the simulated detector has already taken place. This may complicate the path length effect: if the path A is twice the path B for a simulated particle running through a piece of material, then the probability to create a simHit is bigger in case A than in case B. So we have already a correlation between the presence of a simHit and the path length (as simulated with GEANT): that could mean a double counting with the path length effect used to create the map of probabilities. Now, the probability for a particle to create a simHit is already large, close to 1(tbc) while the one to create a HIP is much smaller (~1e-3). For a given increase of the path length, the simHit proba increases by a small amount (since already close to 1) while the HIP proba rescales (almost) linearly. The path length effect is therefore expected to be dominated by the HIP correction.

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by boudoul** [comment]
can somebody retrigger the tests (to hopefully cure the das error)  thanks

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by boudoul** [comment]
Dear @civanch, @mdhildreth, @davidlange6 , could you please help reviewing this PR ? This is important in order to prepare special relvals using Strip ineficiencies - Thanks.
@venturia FYI

**Comment by sdevissc** [comment]
Just one remark about what the current version produces: since the PR there was quite some changes (not pushed yet), especially following a discussion with Andrea, Andreas and Tommaso over the w-e. So The probability map itself should not be considered too seriously for now.

**Comment by civanch** [comment]
@sdevissc , @boudoul , please, confirm that you want our sign (it is not obvious from comments). In general PR is OK....

**Comment by civanch** [comment]
+1

**Comment by sdevissc** [comment]
@civanch Thanks for signing (and sorry for my lack of reply). We will ask for a merge by the end of this week, some changes still have to be pushed, but there are not dramatic at all.

**Comment by slava77** [comment]
@sdevissc 
I wanted to test this with PU35 setting.
Do I just change APVSaturationFromHIP to True,
or do I need to change APVSaturationProbScaling to 1
(from the description, it sounds like the map is derived on PU35, but it's not clear what the default APVSaturationProbScaling setting is)
Please let me know.

Thanks

**Comment by VinInn** [comment]
GSF tracking parameters had never been tuned
http://cmslxr.fnal.gov/lxr/source/TrackingTools/GsfTracking/python/CkfElectronCandidateMaker_cff.py?v=CMSSW_8_1_X_2016-07-06-2300

maybe EGamma POG should have a look and not just set all cuts to ~inf : at some point junk kicks in and efficiency is lost...

**Comment by sdevissc** [comment]
@slava77 
I Slava, 
Wait a bit before doing the test. While the code in SiStripDigitizerAlgo is bascially frozen, the probability map has changed. I'll upload the new version today. And for the parameters, you'll have just to change APVSaturationFromHIP to true (by default it is left to false, while the validation for mass production is not finished)

**Comment by sdevissc** [comment]
Hi all, 

As you may have seen I did a new commit of the map and code. The map was "trained" on the run 274284 (Lumi L1=~6e33), and by playing with the parameter APVSaturationProbScaling I could get the prediction for another inst lumi it is possible to get the hit efficiencies for any other run with lumi L2. The parameter is simply set to L2/L1. 

Here is the global hit efficiency for the condition of run 274284

[globalEfficiencies_274284_s4_s2_45evt_15evt_s7p0_s5.pdf](https://github.com/cms-sw/cmssw/files/355036/globalEfficiencies_274284_s4_s2_45evt_15evt_s7p0_s5.pdf)

and the comparison, after using APVSaturationProbScaling (but of course using the same probability map)

[globalEfficiencies_274956_s7_s5_f0p32.pdf](https://github.com/cms-sw/cmssw/files/355038/globalEfficiencies_274956_s7_s5_f0p32.pdf)

The agreement is generally fairly good but we have to keep in mind a couple of things.

a) No PU reweighting was used to calculate the map, simply because I lack a bit of time.
b) The data points are obtained with ExpressStream, not Prompt-Reco. So to some extent we are comparing apples and guitars, even if the difference expected between the ES and the PR are relatively small (checked in DQM GUI).

To me, the code is now ready for first external (i.e. not by me) comparisons. At least to see if this simulation helps to recover what has been seen by the TRK, BTV, Muon POGs.

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by VinInn** [comment]
@davidlange6 @arizzi @venturia @boudoul 
I strongly advice to integrate it asap and produce high-stat relval for 6x10^33 and 1.2x10^34

**Comment by slava77** [comment]
maybe "do NOT merge now" can be edited out. So that we don't pass a confusing message in release notes.

**Comment by davidlange6** [comment]
Yes, could @sdevissc update the title if this pr is now ready to be considered

**Comment by sdevissc** [comment]
@davidlange6 Done.

**Comment by civanch** [comment]
+1

File: SimTracker/SiStripDigitizer/plugins/SiStripDigitizerAlgorithm.cc [hunk 1]
```diff
@@ -71,6 +77,22 @@ SiStripDigitizerAlgorithm::SiStripDigitizerAlgorithm(const edm::ParameterSet& co
   else LogDebug("SiStripDigitizerAlgorithm")<<" SingleStripNoise: OFF";
   if(CommonModeNoise) LogDebug("SiStripDigitizerAlgorithm")<<" CommonModeNoise: ON";
   else LogDebug("SiStripDigitizerAlgorithm")<<" CommonModeNoise: OFF";
+  
+  std::string line; 
+  APVProbaFile.open((APVProbabilityFile.fullPath()).c_str());
```
Comment (davidlange6) [inline SimTracker/SiStripDigitizer/plugins/SiStripDigitizerAlgorithm.cc]: @sdevissc - could this initialization be done only when needed according to the APVSaturationFromHIP value?

**Comment by davidlange6** [comment]
@iahmad-khan @smuzaffar @degano - could you move 
SimTracker/SiStripDigitizer/data/APVProbaList.txt 
to externals.

File: SimTracker/SiStripDigitizer/plugins/SiStripDigitizerAlgorithm.cc (hunk 1 above)
Comment (sdevissc) [inline SimTracker/SiStripDigitizer/plugins/SiStripDigitizerAlgorithm.cc]: @davidlange6  Right, I'll correct that, thanks!

**Comment by smuzaffar** [comment]
@davidlange6 , sure we should.
I have asked @iahmad-khan  to take care of it.

**Comment by VinInn** [comment]
@lveldere , fyi

**Comment by davidlange6** [comment]
Great. Thanks @iahmad-khan   So the last step is for @sdevissc to remove the file from this pull request. It should be available in the latest ib

**Comment by sdevissc** [comment]
@davidlange6 Ok I'll make that change. Now one question: since this textfile is expected to be changed in the near future (things are still in development), is there any way for me to make changes in this file in the external db? Or do I have to ask each time someone to do it?

Thanks

Simon

**Comment by sdevissc** [comment]
It seems something got strange with the first of the last two commits. But anyway the code seems ok.

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@sdevissc , a general approach for external files: even if you change only one character you provide a new version of the file (with slightly different name) to be integrated into CMSSW and in a new release CMSSW may be switched to use the new file.

**Comment by sdevissc** [comment]
@civanch Ok. But in practice, how do I "push" myself a new version of the textfile to the external db, since it is  (at least for what I understand) not expected to be part of a PR. And does this change anything from the point of view of file handling by the code? Sorry for these very basic questions.

**Comment by civanch** [comment]
+1

**Comment by VinInn** [comment]
is this PR ready to go?
@sdevissc, could you please inform the usual suspects (HIP mailing list) once integrated,
which is the first IB it will be and how to switch on?

thanks

**Comment by VinInn** [comment]
Btw, is it understood if this code will work in PreMixing?

**Comment by davidlange6** [comment]
you can make a pull request to the appropriate repository.  If this sort of data is expected to change with any frequency, the conditions database is the more appropriate home (actually that is true in any case)

https://github.com/cms-data/SimTracker-SiStripDigitizer


On Jul 14, 2016, at 1:13 PM, sdevissc <notifications@github.com<mailto:notifications@github.com>> wrote:


@civanch<https://github.com/civanch> Ok. But in practice, how do I "push" myself a new version of the textfile to the external db, since it is (at least for what I understand) not expected to be part of a PR. And does this change anything from the point of view of file handling by the code? Sorry for these very basic questions.

**Comment by VinInn** [comment]
I would like to stress that we need urgently RelVal-like statistics MC with this PR al "nominal" luminosity and at 1 even 1.2 x10^34.
and then we need to reconstruct those with and w/o Mitigation.
Not sure who is in charge to coordinate all this

**Comment by davidlange6** [comment]
@mdhildreth - could you have a look for @vininn 's question on premixing. Thanks

**Comment by davidlange6** [comment]
ping @mdhildreth 

@vininn, @sdevissc  - what (code) changes are needed to activate this new functionality (false->true in config) and to change its effective lumi (new txt file?)?

**Comment by arizzi** [comment]
@VinInn me/giovanni are collecting the list of samples to produce (I sent an email 1 month ago and we will send an updated reminder now)... but we need this PR to be merged.
Concerning premixing is a second order problem (as for sure it works on the signal tracks)... so can we please go ahead?

**Comment by jlagram** [comment]
The flag is APVSaturationFromHIP in SiStripSimParameters_cfi.py
and the parameter APVSaturationProbScaling for the scaling with the luminosity.

**Comment by boudoul** [comment]
For premixing, assuming this PR is creating troubles with it (I don't know.. @mdhildreth ? ) my worry is that It can be an immediate issue if we backport to 80X ?

**Comment by davidlange6** [comment]
for 81x it seems we are ready - but can someone make a separate PR with the needed customize functions for the relvals samples desired that we can use to do some last tests?

On Jul 19, 2016, at 12:12 PM, arizzi <notifications@github.com<mailto:notifications@github.com>> wrote:


@VinInn<https://github.com/VinInn> me/giovanni are collecting the list of samples to produce (I sent an email 1 month ago and we will send an updated reminder now)... but we need this PR to be merged.
Concerning premixing is a second order problem (as for sure it works on the signal tracks)... so can we please go ahead?

**Comment by gpetruc** [comment]
So, from a customise_commands it seems to me it should be
  `   process.mix.digitizers.strip.APVSaturationFromHIP = True`
and the luminosity can be changed as 
  `   process.mix.digitizers.strip.APVSaturationProbScaling = (Lumi)/(LumiRef)`
where LumiRef is I think about 6E33.

Now running a test on a few events.

Giovanni

**Comment by boudoul** [comment]
ok I can try to prepare a customize function in a spearate PR

**Comment by mdhildreth** [comment]
Hi All,
Sorry, was traveling last week and got behind on git mail.  Premixing will introduce a second-order effect if we do nothing because the dead APVs will be uncorrelated.  There will be an update to the DataMixer code required to implement this in PreMixing, since the final digitization is done there.  More later...

**Comment by arizzi** [comment]
Hi mike do you confirm thay unless we make apvkilled premixed minbias we will juat have the PU fully efficient and only the signal APV killed?

**Comment by gpetruc** [comment]
I've tested in runTheMatrix 25202.0 with the customizes from my previous comment, and I see the expected qualitative behaviour (i.e. decreased number of hits per track when switching it on and increasing the lumi).

@davidlange6: do you want a rebase of this to remove the text file from the history? 

@mdhildreth: so, we can integrate the code as is and it will work for standard mixing (which is what we need for the first relvals), and then we can make a future PR to allow it to work also in premixing?

**Comment by slava77** [comment]
@gpetruc 
in case, from a week ago: see some plots with this PR https://github.com/cms-sw/cmssw/pull/15114#issuecomment-232047911

**Comment by mdhildreth** [comment]
@arizzi: the way the code is currently written, it just randomly kills APVs based on probabilities (and the presence of signal).  So, all of the APV-killing can be done at the digi combination stage in the PreMixing. No need to make APV-killed minbias.  My comment above was based on the previous version which was looking at simHits.

**Comment by mdhildreth** [comment]
@gpetruc: It is easy to integrate the current version so that it works in PreMixing (except I need to know where to get the big text file if it's not in the release).  No changes to this are necessary. I will make a separate PR.

**Comment by mdhildreth** [comment]
@arizzi - sorry, you were correct - I thought the killing block was in the digitizer, not the accumulator.  So, if we want to go with this version, we have to pass through the information that a given APV is killed so that the mixing can kill the correct APV in the final digis.  That's a little tricky, but should be do-able.

Just a question to the list (@gpetruc, @boudoul)  Why not kill on total pulse height, instead of doing it this way?  Isn't that more realistic?  The current algorithm has the same probability of killing an APV if there is 1x10^-6 eV deposited as it does 1 MeV.  [sorry if that question is answered elsewhere]

**Comment by arizzi** [comment]
Inefficiency mostly come from previous BX so it is not (only) correlated
with current event content

Il 19/lug/2016 20:33, "mdhildreth" <notifications@github.com> ha scritto:

**Comment by mdhildreth** [comment]
Yes, but right now, the probability that an APV is killed only depends on how many times one strip is hit anytime during the string of bunch crossings, not how much energy each track deposits.  (I guess you can use the number of Simtracks as a proxy for the energy deposited, but it's quite crude.) The actual calculation of deposited charge is just a multiplicative factor applied at the end, no matter what happens in the meantime.

**Comment by davidlange6** [comment]
The problem is that the relevant time constant is big compared to the time window we simulate.


On 19 Jul 2016, at 20:56, mdhildreth <notifications@github.com<mailto:notifications@github.com>> wrote:


Yes, but right now, the probability that an APV is killed only depends on how many times one strip is hit anytime during the string of bunch crossings, not how much energy each track deposits. (I guess you can use the number of Simtracks as a proxy for the energy deposited, but it's quite crude.) The actual calculation of deposited charge is just a multiplicative factor applied at the end, no matter what happens in the meantime.

**Comment by mdhildreth** [comment]
meaning bigger than 300ns?

**Comment by davidlange6** [comment]
Right.



On 19 Jul 2016, at 21:01, mdhildreth <notifications@github.com<mailto:notifications@github.com>> wrote:


meaning bigger than 300ns?

**Comment by mdhildreth** [comment]
Another question:  can the signal from the track that caused the HIP be lost?  In other words, if a track from the hard-scatter event is the HIP, does that kill its own signal, or do you only loose signal from a previous BX?

**Comment by davidlange6** [comment]
@mdhildreth - as implemented there is no real "track" that caused the HIP. Its purely an ad hoc model that says each APV is affected by a HIP for some fraction of events (maybe the SIM group should learn more about this effect and model on Friday?)

**Comment by boudoul** [comment]
Friday will be tough for the TK DPG  to be at the sim meeting ... : we are in TK week + ICMS ICHEP Approval + TK DPG meeting , so I fear can't split ourselves more than that... but I can provide link to slides presented at the XPOG : 
https://indico.cern.ch/event/544743/contributions/2219208/attachments/1299711/1939698/HIP_XPOG_GB.pdf
and update :+1: https://indico.cern.ch/event/555227/contributions/2241165/attachments/1307820/1955487/XPOG-July-12th.pdf

**Comment by arizzi** [comment]
@mdhildreth just to clarify, this simulation is urgently needed, so we should just verify that is technically ok. We know already that it reproduces what we see in data well enough. 
We can (and should!) on longer term also simulate properly the effect of in-time HIPs as they could be correlated with signal (e.g. in the jets) and it is important to have a model to understand what happens (I did that at test beam time 12 years ago and can provide information on that)... but is not the focus of this PR

**Comment by davidlange6** [comment]
merged as #15242

**Pull request closed without merging**